pandas/tests/reshape/test_concat.py

```diff
@@ -829,12 +831,107 @@ def test_append_preserve_index_name(self):
         assert result.index.name == 'A'
 
+    @pytest.mark.parametrize("df_columns", [
+        pd.RangeIndex(3),
+        pd.CategoricalIndex('A B C'.split()),
+        pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()]),
+        pd.IntervalIndex.from_breaks([0, 1, 2, 3]),
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+        pd.Index([1, 2, 3]),
+    ], ids=lambda x: str(x.dtype))
+    def test_append_same_columns_type(self, df_columns):
+        # GH18359
+
+        # df wider than ser
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=df_columns)
+        ser_index = df_columns[:2]
+        ser = pd.Series([7, 8], index=ser_index, name=2)
+        result = df.append(ser)
+        expected = pd.DataFrame([[1., 2., 3.], [4, 5, 6], [7, 8, np.nan]],
+                                index=[0, 1, 2],
+                                columns=df_columns)
+        assert_frame_equal(result, expected)
+
+        # ser wider than df
+        ser_index = df_columns
+        df_columns = df_columns[:2]
+        df = pd.DataFrame([[1, 2], [4, 5]], columns=df_columns)
+        ser = pd.Series([7, 8, 9], index=ser_index, name=2)
+        result = df.append(ser)
+        expected = pd.DataFrame([[1, 2, np.nan], [4, 5, np.nan], [7, 8, 9]],
+                                index=[0, 1, 2],
+                                columns=ser_index)
+        assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("df_columns, series_index", combinations([
+        pd.RangeIndex(3),
+        pd.Index([4, 5, 6]),
+        pd.Index([7.5, 8.5, 9.5]),
+        pd.CategoricalIndex('A B C'.split()),
+        # pd.CategoricalIndex('A B C'.split(), ordered=True),
+        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
+                          dt.datetime(2013, 1, 3, 6, 10),
+                          dt.datetime(2013, 1, 3, 7, 12)]),
+    ], r=2), ids=lambda x: str(x.dtype))
+    def test_append_different_columns_types(self, df_columns, series_index):
+        # GH18359
+        # see also tests 'test_append_multi_index_raises' and
+        # 'test_append_interval_index_raises' below
+
+        df = pd.DataFrame([[1, 2, 3], [4, 5, 6]], columns=df_columns)
+        ser = pd.Series([7, 8, 9], index=series_index, name=2)
+
+        result = df.append(ser)
+        idx_diff = ser.index.difference(df_columns)
+        combined_columns = Index(df_columns.tolist()).append(idx_diff)
+        expected = pd.DataFrame([[1., 2., 3., np.nan, np.nan, np.nan],
+                                 [4, 5, 6, np.nan, np.nan, np.nan],
+                                 [np.nan, np.nan, np.nan, 7, 8, 9]],
+                                index=[0, 1, 2],
+                                columns=combined_columns)
+        assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize("this_type", [
```

Comment:
you can accomplish this in a better way by making all of these indices into a list and setting it to a variable, then using, call it ``index_dtypes``, then using
``product(index_dtypes, index_dtypes)`` to create the matrix. you can also use this above as well (append_different_column_types), just with ``index_dtypes``

Reply:
I don'y understand you here, sorry, could you expand.

Reply:
```
indexes = [pd.RangeIndex(3),
        pd.Index([4, 5, 6]),
        pd.Index(list("abc")),
        pd.CategoricalIndex('A B C'.split()),
        pd.CategoricalIndex('A B C'.split(), ordered=True),
        pd.IntervalIndex.from_breaks([0, 1, 2, 3]),
        pd.MultiIndex.from_arrays(['A B C'.split(), 'D E F'.split()]),
        pd.DatetimeIndex([dt.datetime(2013, 1, 3, 0, 0),
                          dt.datetime(2013, 1, 3, 6, 10),
                          dt.datetime(2013, 1, 3, 7, 12)])]
@pytest.fixture(params=indexes)
def this_type(request):
    return request.param

@pytest.fixture(params=indexes)
def that_type(params=indexes):
    return request.param
.....

def append_different_coumns_types_raises(self, this_type, that_type):
   ....
```